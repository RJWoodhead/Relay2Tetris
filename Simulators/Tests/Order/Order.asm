// Test order of operations and usage of AM register
//
// 1) Test what happens when A and M are updated in same instruction
//
D=-1        // D=-1
@0          // A=0
M=D         // RAM[0] = -1
@1          // A=1
M=D         // RAM[1] = -1
@2          // A=2
M=D         // RAM[2] = -1
A=0         // A=0
D=1         // D=1
AM=D        // A=1, RAM[0] = 1 or RAM[1] = 1? Answer: RAM[0] = 1
//
// 2) Test what happens when A is updated and a branch occurs
//
@GOOD       // A=Address of (GOOD)
D=A         // D=Address of (GOOD)
@BAD        // A=Address of (BAD)
A=D;JMP     // A=Address of (GOOD), JMP executes
//
(GOOD)      // If jump occurs using updated A, we get here
@2          // A=2
M=1         // RAM[2]=1, then fall through
//
(BAD)       // If we jump directly here, RAM[2] will be -1
//
// Tighter and detectable-in-hardware end loop
//
@END        // End program
(END)
0;JMP
//
// Result: RAM[0] = 1, RAM[1] = -1, RAM[2] = 1
//
