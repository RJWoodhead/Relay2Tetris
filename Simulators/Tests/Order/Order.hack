1110111010010000
0000000000000000
1110001100001000
0000000000000001
1110001100001000
0000000000000010
1110001100001000
1110101010100000
1110111111010000
1110001100101000
0000000000001110
1110110000010000
0000000000010000
1110001100100111
0000000000000010
1110111111001000
0000000000010001
1110101010000111
