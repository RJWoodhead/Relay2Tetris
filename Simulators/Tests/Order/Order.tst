// A Register order of operations test

load Order.asm,
output-file Order.out,
compare-to Order.cmp,
output-list RAM[0]%D1.6.1 RAM[1]%D1.6.1 RAM[2]%D1.6.1;

repeat 50 {
  ticktock;
}

output;
