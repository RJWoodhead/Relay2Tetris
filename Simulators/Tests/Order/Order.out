| RAM[0] | RAM[1] | RAM[2] |
|      1 |     -1 |      1 |
